docs: update "Topology-Aware Provisioning" (#1427)
longhorn/website: content/docs/1.13.0/nodes-and-volumes/nodes/topology-aware-provisioning.md

weight: 6
---

Topology-aware provisioning allows you to align Kubernetes pod placement and Longhorn data placement with the topology of your cluster. In multi-zone or multi-region clusters, this helps avoid permanent cross-zone or cross-region I/O when a pod is constrained to one failure domain but the replicas or erasure-coded shards backing the volume are scheduled somewhere else.

Longhorn has two levels of topology control:

- **PV topology** controls the `nodeAffinity` rules that Kubernetes writes into a PersistentVolume (PV). This determines where pods using the PV can run.
- **Volume topology** controls where Longhorn stores the volume data. This determines where replicas, rebuild replicas, auto-balance candidates, and erasure-coded shards can be placed.

> By default, Longhorn keeps the existing behavior: StorageClasses without topology settings are unconstrained. To make a volume topology-aware, opt in with the `volumeTopology` StorageClass parameter.

Reviewer: ```suggestion
Topology-aware provisioning allows you to align Kubernetes pod placement and Longhorn data placement with your cluster's topology. In multi-zone or multi-region clusters, this prevents permanent cross-zone or cross-region I/O that occurs when a pod is constrained to one failure domain while its backing data (replicas or erasure-coded shards) is scheduled in another.

Longhorn provides two levels of topology control:

- **PV topology**: Controls the `nodeAffinity` rules Kubernetes writes into a PersistentVolume (PV), dictating where pods using the PV can run.
- **Volume topology**: Controls where Longhorn stores volume data, dictating the placement of replicas, rebuild replicas, auto-balance candidates, and erasure-coded shards.

> **Note**: By default, StorageClasses without topology settings are unconstrained. To make a volume topology-aware, opt in using the `volumeTopology` StorageClass parameter.
```

   ```shell
   kubectl get nodes --label-columns topology.kubernetes.io/zone,topology.kubernetes.io/region
   ```

2. Configure the **CSI Allowed Topology Keys** setting in Longhorn. Set the value to a comma-separated list of topology keys that Longhorn should use when building PV `nodeAffinity`.

   - **Longhorn UI**: Go to **Setting > General > CSI Allowed Topology Keys** and enter, for example, `topology.kubernetes.io/zone` or `topology.kubernetes.io/zone,topology.kubernetes.io/region`.
   - **Longhorn API / kubectl**:
     ```shell
     kubectl -n longhorn-system edit settings.longhorn.io csi-allowed-topology-keys
     ```
     Set the `value` field to the topology keys you want Kubernetes PVs to use.

   > **Note:** After changing this setting, manually restart the `longhorn-csi-plugin` DaemonSet for the change to take effect. Topology is applied correctly only after the CSI plugin pod on each node has restarted.

3. For topology-aware dynamic provisioning, set the StorageClass `volumeBindingMode` to `WaitForFirstConsumer`. This lets Kubernetes choose a consumer pod's node before Longhorn resolves the volume topology.

## How It Works

> The following flowchart shows how a topology-aware volume is provisioned, from PVC creation to where the resolved failure domain is enforced:

Reviewer: ```suggestion
The following flowchart illustrates the provisioning lifecycle of a topology-aware volume, from PVC creation to failure domain enforcement:
```

    B --> C["CSI CreateVolume receives<br/>the selected node's topology"]
    C --> D["Keep only the csi-allowed-topology-keys<br/>→ sets the PV nodeAffinity"]
    D --> E{"volumeTopology?"}
    E -->|any / unset| F["No constraint stored<br/>Replicas scheduled as before"]
    E -->|zonal| G["Pin the volume to one zone<br/>Force replicaZoneSoftAntiAffinity = enabled"]
    E -->|regional| H["Pin the volume to one region<br/>Replicas may still span zones in that region"]
    G --> I["Constraint applies to every placement:<br/>new replicas, rebuilds, replica-count changes,<br/>auto-balance, and erasure-coded shards.<br/>No fallback to another zone or region."]
    H --> I
```

When a PVC is created against a StorageClass that uses the Longhorn CSI driver, Kubernetes sends Longhorn a list of accessible topologies. Longhorn first filters those topology segments using `csi-allowed-topology-keys`. The filtered topology is used for the PV `nodeAffinity`.

If `volumeTopology` is unset or set to `any`, Longhorn does not store a volume placement constraint. Replica scheduling behaves as it did before.

> If `volumeTopology` is set to `zonal` or `regional`, Longhorn resolves one failure domain at creation time and stores it in the Volume spec as `topologyRequirement`. Longhorn then returns that same resolved topology to Kubernetes so the PV `nodeAffinity` and Longhorn data placement refer to the same domain. The stored requirement is immutable and is enforced for initial replica placement, rebuilds, replica count changes, replica auto-balance, and erasure-coded shard placement. Longhorn does not fall back to a different zone or region when the resolved domain has no capacity.

Reviewer: ```suggestion
The provisioning process follows these steps:

1.  **Topology Discovery**: When a PVC is created against a StorageClass using the Longhorn CSI driver, Kubernetes sends Longhorn a list of accessible topologies. 
2.  **Filtering**: Longhorn filters these topology segments based on `csi-allowed-topology-keys` and uses the result to set the PV `nodeAffinity`.
3.  **Volume Topology Enforcement**: Longhorn determines data placement constraints based on the `volumeTopology` setting:
    *   **Unset or `any`**: No placement constraint is stored. Replica scheduling behaves normally.
    *   **`zonal` or `regional`**: Longhorn resolves a single failure domain, stores it in the Volume spec as `topologyRequirement`, and returns it to Kubernetes. This ensures both PV `nodeAffinity` and Longhorn data placement are bound to the same domain.

> **Warning**: A stored `topologyRequirement` is immutable. Longhorn strictly enforces this domain for initial placement, rebuilds, replica count changes, auto-balancing, and erasure-coded shards. If the resolved domain has no capacity, Longhorn **will not** fall back to another zone or region.
```

- `volumeTopology: "zonal"` and `volumeTopology: "regional"` resolve one failure domain at creation time and keep the PV and Longhorn data placement aligned for the lifetime of the volume.
- If nodes have the requested topology label but `csi-allowed-topology-keys` does not include that key, provisioning with `volumeTopology: "zonal"` or `"regional"` is rejected. Add the required key to the setting and restart the CSI plugin.
- If nodes do not report the requested topology label at all, Longhorn creates the volume without a topology constraint. This allows one StorageClass to work on both labeled multi-zone clusters and unlabeled single-zone clusters.

## Notes and Warnings

- Do **not** use `allowedTopologies` together with `dataLocality: strict-local`. The PV `nodeAffinity` is immutable once set and will conflict with Longhorn's strict-local volume pinning. See [Data Locality](../../../high-availability/data-locality/) for details.
- Do **not** use `volumeTopology: "zonal"` together with `replicaZoneSoftAntiAffinity: "disabled"`. Longhorn rejects this combination because all replicas must stay in one zone.

> - For zonal or regional volume placement, prefer `volumeBindingMode: WaitForFirstConsumer` so the resolved failure domain follows the first consumer pod's scheduling decision.

Reviewer 2: It may be worth stating what `Immediate` binding does, since "the volume picks the domain and the pod follows" is the surprising part for users coming from `any` — and this is probably also the item for the v1.13.0 important notes (the issue is labeled `require/important-note`).

```suggestion
- For zonal or regional volume placement, prefer `volumeBindingMode: WaitForFirstConsumer` so the resolved failure domain follows the first consumer pod's scheduling decision.
- With `Immediate` binding, the failure domain is resolved from the provisioner-supplied topology at creation time (effectively round-robin across domains), and the first consumer pod is then constrained to that domain through the PV `nodeAffinity`.
```

> - Existing volumes and StorageClasses are unchanged unless you set `volumeTopology` on a StorageClass used for new PVCs.

Reviewer 2: One interaction worth a note here: a linked clone is always placed next to the source replica, so its data placement is already fixed by the source volume. With `volumeTopology: "zonal"` on the clone's StorageClass, the clone can be scheduled only when the first consumer pod happens to land in the source replica's zone — otherwise the two hard constraints cannot both be satisfied and the volume stays unschedulable.

```suggestion
- Existing volumes and StorageClasses are unchanged unless you set `volumeTopology` on a StorageClass used for new PVCs.
- Avoid setting `volumeTopology` on a StorageClass used for linked clones. A linked clone is always placed next to the source replica, so its data placement is already fixed; a `zonal` clone volume can be scheduled only when the first consumer pod lands in the source replica's zone, and stays unschedulable otherwise.
```
